File: src/main/java/ru/citeck/ecos/process/domain/procdef/service/ProcDefService.java
package ru.citeck.ecos.process.domain.procdef.service;

import ru.citeck.ecos.apps.artifact.ArtifactRef;
import ru.citeck.ecos.process.domain.proc.dto.NewProcessDefDto;
import ru.citeck.ecos.process.domain.procdef.dto.ProcDefWithDataDto;
import ru.citeck.ecos.process.domain.procdef.dto.ProcDefDto;
import ru.citeck.ecos.process.domain.procdef.dto.ProcDefRevDto;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records2.predicate.model.Predicate;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface ProcDefService {

    ProcDefDto uploadProcDef(NewProcessDefDto processDef);

    ProcDefDto uploadNewRev(ProcDefWithDataDto dto);

    List<ProcDefWithDataDto> findAllWithData(Predicate predicate, int max, int skip);

    List<ProcDefDto> findAll(Predicate predicate, int max, int skip);

    long getCount();

    long getCount(Predicate predicate);

    String getCacheKey();

    Optional<ProcDefRevDto> getProcessDefRev(String procType, UUID procDefRevId);

    Optional<ProcDefRevDto> findProcDef(String procType, RecordRef ecosTypeRef, List<String> altTypes);

    Optional<ProcDefWithDataDto> getProcessDefById(ArtifactRef id);
}

File: src/main/java/ru/citeck/ecos/process/domain/procdef/service/ProcDefService.kt
package ru.citeck.ecos.process.domain.procdef.service

import ru.citeck.ecos.process.domain.proc.dto.NewProcessDefDto
import ru.citeck.ecos.process.domain.procdef.dto.ProcDefDto
import ru.citeck.ecos.process.domain.procdef.dto.ProcDefRef
import ru.citeck.ecos.process.domain.procdef.dto.ProcDefRevDto
import ru.citeck.ecos.process.domain.procdef.dto.ProcDefWithDataDto
import ru.citeck.ecos.records2.predicate.model.Predicate
import ru.citeck.ecos.webapp.api.entity.EntityRef
import java.util.*

interface ProcDefService {

    fun uploadProcDef(processDef: NewProcessDefDto): ProcDefDto

    fun uploadProcDefDraft(processDef: NewProcessDefDto): ProcDefDto

    fun uploadNewRev(dto: ProcDefWithDataDto, comment: String = ""): ProcDefDto

    fun uploadNewDraftRev(dto: ProcDefWithDataDto, comment: String = "", forceMajorVersion: Boolean = false): ProcDefDto

    fun findAllWithData(workspaces: List<String>, predicate: Predicate?, max: Int, skip: Int): List<ProcDefWithDataDto>

    fun findAll(workspaces: List<String>, predicate: Predicate, max: Int, skip: Int): List<ProcDefDto>

    fun getCount(workspaces: List<String>): Long

    fun getCount(workspaces: List<String>, predicate: Predicate): Long

    fun getCacheKey(): String

    fun getProcessDefRev(procType: String, procDefRevId: UUID): ProcDefRevDto?

    fun saveProcessDefRevDeploymentId(procDefRevId: UUID, deploymentId: String)

    fun getProcessDefRevs(procDefRevIds: List<UUID>): List<ProcDefRevDto>

    fun getProcessDefRevs(ref: ProcDefRef): List<ProcDefRevDto>

    fun getProcessDefRevByDeploymentId(deploymentId: String): ProcDefRevDto?

    fun getProcessDefRevByDeploymentIds(deploymentIds: List<String>): List<ProcDefRevDto>

    /**
     * return first enabled process definition by ecosTypeRef hierarchy
     */
    fun findProcDef(
        procType: String,
        workspace: String,
        ecosTypeRef: EntityRef?,
        alfTypes: List<String>?
    ): ProcDefRevDto?

    fun getProcessDefById(id: ProcDefRef): ProcDefWithDataDto?

    fun delete(ref: ProcDefRef)

    fun listenChanges(type: String, action: (ProcDefDto) -> Unit)
}
